versioned_hdf5/subchunk_map.py

```diff
@@ -17,308 +22,678 @@
+        )
 
-    can be translated into
+    @cython.ccall
+    def chunk_submap(
+        self,
+        chunk_start_idx: Py_ssize_t,
+        chunk_stop_idx: Py_ssize_t,
+        shift: bint,
+    ) -> tuple[NDArray[np.intp], NDArray[np.intp]]:
+        start, stop = self.chunk_start_stop(chunk_start_idx, chunk_stop_idx)
+
+        sub = np.flatnonzero(self.idx[start:stop])
+        if len(sub) == 0:
+            raise ChunkMapIndexError()
+        if not shift:
+            sub += start
+
+        offsets = self._chunk_selected_offsets
+        counts = self._chunk_selected_counts
+        n_before = offsets[chunk_start_idx]
+        n_select = 0
+        for i in range(chunk_start_idx, chunk_stop_idx):
+            n_select += counts[i]
+        n_after = offsets[len(offsets) - 1] - offsets[chunk_stop_idx]
+
+        return (
+            # FIXME Why does the test fail if this is replaced with
+            #   slice(n_before, n_before + n_select, 1) ?
+            np.concatenate(
+                [
+                    np.zeros(n_before, dtype=bool),
+                    np.ones(n_select, dtype=bool),
+                    np.zeros(n_after, dtype=bool),
+                ]
+            ),
+            sub,
+        )
 
-    >> for chunk, arr_idx_raw, index_raw in as_subchunk_map(ds.chunk_size, idx, ds.shape):
-    ..     ds.data_dict[chunk][index_raw] = arr[arr_idx_raw]
 
-    :param chunk_size: the `ChunkSize` of the Dataset
-    :param idx: the "index" to read from / write to the Dataset
-    :param shape: the shape of the Dataset
-    :return: a generator of `(chunk, arr_idx_raw, index_raw)` tuples
+@cython.cclass
+class IntegerMapper(IndexChunkMapper):
+    idx: Py_ssize_t
+
+    def __init__(
+        self,
+        idx: Py_ssize_t,
+        chunk_size: Py_ssize_t,
+        dset_size: Py_ssize_t,
+    ):
+        assert 0 <= idx < dset_size
+        self.idx = idx
+        chunk_indices = np.array([idx // chunk_size], dtype=np.intp)
+        super().__init__(chunk_indices, chunk_size, dset_size)
+
+    @cython.ccall
+    def chunks_indexer(self):
+        # a[i] would change the shape
+        # a[[i]] (the default without overriding this method) would return a copy
+        # a[i:i+1] returns a view
+        i = self.chunk_indices[0]
+        return slice(i, i + 1, 1)
+
+    @cython.ccall
+    def whole_chunks_indexer(self):
+        if self.chunk_size == 1:
+            return self.chunks_indexer()
+        else:
+            return slice(0, 0, 1)
+
+    @cython.ccall
+    def chunk_submap(
+        self,
+        chunk_start_idx: Py_ssize_t,
+        chunk_stop_idx: Py_ssize_t,
+        shift: bint,
+    ) -> tuple[int, int]:
+        # Raise if we're selecting an index out of bounds
+        start, _ = self.chunk_start_stop(chunk_start_idx, chunk_stop_idx)
+        return self.idx, self.idx - start if shift else self.idx
+
+
+@cython.cclass
+class EverythingMapper(IndexChunkMapper):
+    """Select all points along an axis [:]"""
```

Comment:
Tecnically the same as `SliceMapper(0, dset_size, 1, chunk_size, dset_size)` but much cleaner and faster